### ISSUE TITLE
Encoding question

### ISSUE THREAD

**Issue by t-dim** [msoos/cryptominisat#720]
Dear Mate hi, 

I know Cryptominisat can handle a mix of CNF and XOR clauses, but what
about formulas whose clauses contain sums (OR's) of XOR's like the
example below?

f = (x1 XOR x2 + ~x1 XOR x3 + x4 XOR x2) AND (~x1 XOR x4 + ~x3 + x3 XOR x4 + x1 XOR x4) AND (...more clauses of this kind...)

So, basically, each clause in f is the disjunction of XOR groups (where an XOR group may consist of 1 or more standard literals). In the example above, the first clause is the disjunction of 3 XOR groups consisting of 2 literals each. But typically an XOR group may have as many as N/2 literals on average, where N is the number of variables in the formula.

This can be translated this to a mix of usual CNFs and XORs by introducing extra variables (a la Tseitin), so the first clause would have become:

(y1 + y2 + y3) AND (y1 <=>x1 XOR x2) AND (y2<=> ~x1 XOR x3) AND (y3<=> x4 XOR x2) 

where the equivalences are basically plain XOR clauses. I tried this approach already but unfortunately cryptominisat takes an extremely large amount of time even with formulas with 400-500 variables. As a comparison, cryptominisat on random 3-SAT formulas with 400-500 vars takes just a few milliseconds. But in my case, the formulas are disjunctions of XORs groups with about 200 vars each on average...

Any help would be appreciated. My feeling is that these formulas are very hard to solve in general, so dealing with those efficiently may also advance the state of the art in SAT solving. I can also provide a sample of formulas if needed..

Thanks and best regards,

Tassos

**Comment by msoos** [comment]
Hi,

Encoding `(x1 XOR x2 + ~x1 XOR x3 + x4 XOR x2)` is best done with Tseitin encoding as you mentioned. This should be an okay encoding. However, knowing the underlying problem may help in designing a better encoding. I am guessing the underlying problem is not just a bunch of XORs, but something different, like e.g. some circuit, or some scheduling problem, or some math problem. Once I know that, it may be possible that there could be other encodings that are better. As-is, that's the best.

Regarding speed, the size of the formula, or CryptoMiniSat's speed on random CNFs is not related to solving speed of a specific formula. It's more about its structure than anything else, as you correctly suspected :)

If you provide an example, I can check some simple things. However, re-engineering the encoding may be a long-term better approach.

I hope the above helps,

Mate

**Comment by t-dim** [comment]
Dear Mate, thanks for the quick reply.

Here's the whole story. Formulas of this kind arise in the context of an obfuscator for SAT formulas we have built. The idea is to obfuscate a SAT instance prior to outsourcing it to the cloud for solving it... For example, I may have a hard formula f that encodes an industrial instance and I want to outsource it to you for solving it for a fee. However, outsourcing f may reveal considerable information about the underlying instance, so instead I outsource an obfuscated version obf(f) of it. The main property is that Obf(f) is satisfiable IF AND ONLY IF the original formula is satisfiable. Furthermore, from an assignment of obf(f), I can recover easily an assignment of f.

Hence the obfuscator creates a formula which is in some sense "equivalent" to the original one but "disguised" so that no information about the original one can leak. This concept has been properly formalized ...

In any case, the obfuscator creates formulas of the kind I described.. The question is how much harder are the obfuscated formulas from the original ones... I attach two obfuscated instances (both SATISFIABLE) resulting from random 3SAT formulas on 200 and 300 vars, respectively. The first can be solved (in my machine, running cryptominisat 5 with default parameters) in about 390 seconds, while the second was INDETERMINATE after 15000 seconds.  
[_The increased number of variables and clauses shown in the files is due to the addition of extra variables (recall Tseitin) for every XOR group encountered in each clause_.]
[instances.zip](https://github.com/msoos/cryptominisat/files/11471163/instances.zip)

Your feedback is welcome.
